File: HMRepeatableTapGestureRecognizer/Class/HMRepeatableTapGestureRecognizer.m
//
//  HMRepeatableTapGestureRecognizer.m
//  HMRepeatableTapGestureRecognizer
//
//  Created by Muronaka Hiroaki on 2015/09/03.
//  Copyright (c) 2015年 Muronaka Hiroaki. All rights reserved.
//

#import <UIKit/UIGestureRecognizerSubclass.h>
#import "HMRepeatableTapGestureRecognizer.h"
#import <HMTargetActionListFramework/HMTargetActionListFramework.h>

@interface HMRepeatableTapGestureRecognizer()

@property(nonatomic, strong) NSTimer* longPressTimer;
@property(nonatomic, strong) NSTimer* repeatActionTimer;
@property(nonatomic, strong) HMTargetActionList* targetActionList;
@property(nonatomic, assign) NSInteger currentTouchesCount;

@end

@implementation HMRepeatableTapGestureRecognizer

-(instancetype)initWithTarget:(id)target action:(SEL)action {
    
    self = [super initWithTarget:self action:@selector(handleFromSuper:)];
    
    if( self ) {
        self.targetActionList = [HMTargetActionList new];
        [self.targetActionList addTarget:target action:action];
        
        _minimumPressDuration = 0.5;
        _repeatInterval = 0.2;
        _numberOfTouchesRequired = 1;
        _currentTouchesCount = 0;
    }
    
    return self;
}

/////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark -
#pragma mark action

-(void)handleFromSuper:(UIGestureRecognizer*)gesture {
    switch(gesture.state) {
        case UIGestureRecognizerStateBegan:
        case UIGestureRecognizerStateFailed:
        case UIGestureRecognizerStatePossible:
        case UIGestureRecognizerStateCancelled:
            [self.targetActionList fireWithSender:self];
            break;
        case UIGestureRecognizerStateEnded:
            if( self.currentTouchesCount == self.numberOfTouchesRequired ) {
//                [self.targetActionList fireWithSender:self];
            }
            break;
        case UIGestureRecognizerStateChanged:
            break;
        default:
            break;
    }
}

/////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark -
#pragma mark override methods

-(void)addTarget:(id)target action:(SEL)action {
    [self.targetActionList addTarget:target action:action];
}

-(void)removeTarget:(id)target action:(SEL)action {
    [self.targetActionList removeTaget:target action:action];
}

-(void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    self.currentTouchesCount += touches.count;
    
    if( self.currentTouchesCount < self.numberOfTouchesRequired ) {
        return;
    } else if( self.currentTouchesCount > self.numberOfTouchesRequired ) {
        if ( self.state == UIGestureRecognizerStatePossible ) {
            self.state = UIGestureRecognizerStateFailed;
        } else {
            self.state = UIGestureRecognizerStateCancelled;
        }
        return;
    }
    
    [super touchesBegan:touches withEvent:event];
    self.state = UIGestureRecognizerStateBegan;
    
    if( self.minimumPressDuration > 0.0 ) {
        self.longPressTimer = [NSTimer scheduledTimerWithTimeInterval:self.minimumPressDuration target:self selector:@selector(startRepeatAction:) userInfo:nil repeats:NO];
    } else {
        [self startRepeatAction:nil];
    }
}

-(void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
    [super touchesEnded:touches withEvent:event];
    self.state = UIGestureRecognizerStateEnded;
}

-(void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event {
    [super touchesMoved:touches withEvent:event];
}

-(void)touchesCancelled:(NSSet *)touches withEvent:(UIEvent *)event {
    [super touchesCancelled:touches withEvent:event];
}

-(void)reset {
    self.currentTouchesCount = 0;
    [self cancelTimers];
    [super reset];
    
}

/////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark -
#pragma mark timer

-(void)startRepeatAction:(NSTimer*)timer {
    if( self.repeatInterval > 0.0 ) {
        self.repeatActionTimer = [NSTimer scheduledTimerWithTimeInterval:self.repeatInterval target:self selector:@selector(repeatAction:) userInfo:nil repeats:YES];
    }
}

-(void)repeatAction:(NSTimer*)timer {
    self.state = UIGestureRecognizerStateChanged;
    [self.targetActionList fireWithSender:self];
}

-(void)cancelTimers {
    [self.longPressTimer invalidate];
    self.longPressTimer = nil;
    
    [self.repeatActionTimer invalidate];
    self.repeatActionTimer = nil;
}

@end

File: Sources/HMRepeatableTapGestureRecognizer/HMRepeatableTapGestureRecognizer.swift
//
//  HMRepeatableTapGestureRecognizer.swift
//
//  Created by Hiroaki Muronaka on 2021/10/08.
//
import UIKit
import UIKit.UIGestureRecognizerSubclass
import HMTargetActionList

/// Viewを押し続けると連続して.changeイベントを発火するGestureRecognizer
///
/// USAGE
///
/// ```swift
/// let gesture = HMRepeatableGestureRecognizer() { _ in
///    // NOTE gesture.stateは参照しないこと。
///    NSLog("event fired")
/// }
/// self.button.addGesture(gesture)
/// ```
///
/// UIGestureRecognizerはtouchMovedをoverrideしたとしてもtouchMovedが発生したときに.changedイベントが発火されるため、
/// UIGestureRecognizer.stateは利用しない実装としている。
open class HMRepeatableTapGestureRecognizer: UIGestureRecognizer {
    
    //MARK: public attributes
    
    public typealias Action = (HMRepeatableTapGestureRecognizer) -> ()
    
    /// リピートが始まるまでの間隔(seconds)
    public let minimumPressDuration: TimeInterval
    
    /// リピート間隔(seconds)
    public let repeatInterval: TimeInterval
    
    public let numberOfTouchsRequired: Int
    
    /// 発火するアクション
    public let action: Action?
    
    //MARK: private attributes
    /// リピートするまでの待機用タイマ
    private var longPressTimer: Timer?
    
    /// リピートタイマ
    private var repeatActionTimer: Timer?
    
    /// タッチ数
    private var currentTouchsCount = 0
    
    private var targetActionList: HMTargetActionList = HMTargetActionList()
    
    //MARK: constructors
    
    /// gestureを生成する
    /// - Parameters:
    ///   - minimumPressDuration: minimumPressDuration description
    ///   - repeatInterval: repeatInterval description
    ///   - numberOfTouchRequired: numberOfTouchRequired description
    ///   - action: action description
    public init(minimumPressDuration: TimeInterval = 0.5, repeatInterval: TimeInterval = 0.2, numberOfTouchRequired: Int = 1, action: Action? = nil) {
        
        self.minimumPressDuration = minimumPressDuration
        self.repeatInterval = repeatInterval
        self.numberOfTouchsRequired = numberOfTouchRequired
        self.action = action
        
        super.init(target: nil, action: nil)
    }
}

//MARK: implementations
extension HMRepeatableTapGestureRecognizer {
   
    private func cancelTimers() {
        self.repeatActionTimer?.invalidate()
        self.longPressTimer?.invalidate()
        self.repeatActionTimer = nil
        self.longPressTimer = nil
    }
    
    private func startRepeatTimer() {
        guard self.repeatInterval > 0 else {
            return
        }
        self.repeatActionTimer = Timer.scheduledTimer(withTimeInterval: self.repeatInterval, repeats: true, block: { _ in
            self.fireAction()
        })
    }
    
    private func fireAction() {
        self.state = .changed
        self.action?(self)
        self.targetActionList.fire(self)
    }
}
//MARK: override methods
extension HMRepeatableTapGestureRecognizer {
    
    open override func addTarget(_ target: Any, action: Selector) {
        self.targetActionList.addTarget(target as AnyObject, action: action)
    }
    
    open override func removeTarget(_ target: Any?, action: Selector?) {
        if let target = target {
            if let action = action {
                self.targetActionList.removeTarget(target as AnyObject, action: action)
            } else {
                self.targetActionList.removeActions(target as AnyObject)
            }
        }
    }
    
    open override func reset() {
        self.currentTouchsCount = 0
        self.cancelTimers()
        super.reset()
    }
     
    open override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent) {
        self.currentTouchsCount += touches.count
        guard self.currentTouchsCount == self.numberOfTouchsRequired else {
            if self.currentTouchsCount > self.numberOfTouchsRequired {
                self.state = (self.state == .possible ? .failed : .cancelled)
            }
            return
        }
        
        super.touchesBegan(touches, with: event)
        
        if self.minimumPressDuration > 0 {
            self.longPressTimer = Timer.scheduledTimer(withTimeInterval: self.minimumPressDuration, repeats: false, block: { _ in
                self.startRepeatTimer()
            })
        } else {
            self.startRepeatTimer()
        }
        self.state = .began
        self.fireAction()
    }
    
    open override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent) {
        super.touchesCancelled(touches, with: event)
        self.state = .cancelled
    }
    
    open override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent) {
        super.touchesEnded(touches, with: event)
        self.state = .ended
    }
   
}
